packages/token-search-discovery-controller/src/token-search-discovery-controller.ts

```diff
@@ -0,0 +1,122 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseController } from '@metamask/base-controller';
+import type { TokenSearchResponseItem } from './types';
+
+export const TOKENSEARCH_EVENTS = {
+  SEARCH_COMPLETED: 'SEARCH_COMPLETED',
+} as const;
+
+const name = 'TokenSearchDiscoveryController';
+
+type Events = {
+  type: (typeof TOKENSEARCH_EVENTS)[keyof typeof TOKENSEARCH_EVENTS];
+  payload: [TokenSearchResponseItem[]];
+};
+
+export type TokenSearchDiscoveryControllerMessenger =
+  RestrictedControllerMessenger<
+    typeof name,
+    never,
+    Events,
+    never,
+    (typeof TOKENSEARCH_EVENTS)[keyof typeof TOKENSEARCH_EVENTS]
+  >;
+
+export type TokenSearchDiscoveryState = {
+  recentSearches: TokenSearchResponseItem[];
+  lastSearchTimestamp: number | null;
+};
+
+const defaultState: TokenSearchDiscoveryState = {
+  recentSearches: [],
+  lastSearchTimestamp: null,
+};
+
+export class TokenSearchDiscoveryController extends BaseController<
+  typeof name,
+  TokenSearchDiscoveryState,
+  TokenSearchDiscoveryControllerMessenger
+> {
+  private readonly baseUrl: string;
```

Comment:
This is not documented anywhere, however, we use the ES private field syntax in this repo instead of TypeScript's `private` keyword, because it truly hides code instead of "soft hiding" it. More here: https://www.typescriptlang.org/docs/handbook/2/classes.html#caveats

``` suggestion
  readonly #baseUrl: string;